first draft of advice for authorization perf optimization (#123)
biscuit-auth/website: docs/src/recipes/datalog-performance.md

## Datalog performance contributors

As stated above, there are a lot of external factors that contribute to the final time and resource costs of the authorization process.

Other things being equal, some elements in datalog code tend to have a disproportionate effect on performance. Please don't apply them without proper measurements.

With biscuit-rust, you can see how much time was spent evaluating datalog in an authorizer with `Authorizer.execution_time()`. This does not replace performance measurements, but can give you a simple way to compare datalog snippets. Authorizer snapshots carry this information and can be inspected with biscuit-cli through `biscuit inspect-snapshot` or with the web inspector.

### Number of rules

The number of rules is a direct contributor to evaluation performance. The datalog engine tries to match every rule with every fact to produce new facts, and then tries again with new facts until no new facts are produced.

- authorization contexts with a lot of rules will take more time to compute

> - rules generating facts matched by other rules with will require more iterations before convergence

Author: le with est en trop

> - rules generating facts matched by other rules with will require more iterations before convergence

Author: ```suggestion
- rules generating facts matched by other rule will require more iterations before convergence
```

### Expression evaluation

> This part is implementation-dependent, advice applies primarily to the rust implementation.

> Rules can contain expressions, that are evaluated after facts are matched. The biscuit specifications describes an evaluation strategy based on a stack machine, which aims at providing a fast evaluation.

Author: les expressions

> Rules can contain expressions, that are evaluated after facts are matched. The biscuit specifications describes an evaluation strategy based on a stack machine, which aims at providing a fast evaluation.

Author: ```suggestion
Rules can contain expressions, that are evaluated after facts are matched. The biscuit specification describes an evaluation strategy based on a stack machine, which aims at providing a fast evaluation.
```

# Authorization performance

> Authorization is likely part of your request handling hot path. As such, it is natural to try and make it as fast as possible. Signature verification makes up for an important part of the whole authorization process.

Author: c'est une étape qui prend du temps. significant serait peut-être mieux

## Authorization process breakdown

The authorization process can be broken down in 4 parts:

- parsing;
- signature verification;
- datalog generation;
- datalog evaluation.

> Parsing is typically not a big contributor and depends only on the token size. Signature verification is an important contributor and depends mostly on the number of blocks (and their size). Datalog generation and datalog evaluation happen in tandem. That's the part where you have the most leverage. Datalog generation purely depends on how your application is designed. In many cases it can be done statically and thus have a negligible contribution to the overall runtime. Datalog evaluation depends on the actual datalog code that is evaluated.

Author: ça serait bien de le repréciser. j'ai retouché le texte déjà écrit du coup l'ordre des concepts et des termes lors de l'écriture n'est pas celui du document. Si ça se trouve je fais référence à une utilisation de "contributor" qui est plus loin dans le texte

## The first rule of performance optimization

> Don't do it

> Benchmarks done with biscuit-rust show that the whole process (parsing, signatures verification and datalog evaluation) usually clocks in at around one millisecond. In a lot of cases, that will not be a bottleneck and thus not where you should work on performance optimization.

Reviewer: Plus haut tu mets "signature verification" (au singulier donc), faudrait harmoniser je pense

Reviewer: Aussi, précédemment tu cites 4 étapes, mais ici en parlant du "whole process" tu n'en cites que 3. Pourquoi sauter le datalog generation ici ?

Author: parceque j'ai introduit la génération de datalog en étape séparée après avoir écrit ce bout de texte ^^"

## Measure

> When it comes to performance optimization, the first step is always to measure things. First to determine if optimization is even needed, then to quantify progress.

Reviewer: "measure things": qu'est-ce qu'on mesure ? Est-ce que tu fais une différence entre mesurer et analyser (utilisé plus bas)

Author: on mesure le temps d'exécution de chaque étape. c'est une étape qui doit être faite en amont (outiller le code), pour pouvoir être capables d'analyser les résultats par la suite

> 

Reviewer: Saut de ligne à enlever imo. Les deux paragraphes vont ensemble.

This part entirely depends on your tech stack. You can start with coarse-grained traces telling you how long the whole authorization process takes, and then only dig down if optimization is needed.

> Then, there are two steps in the authorization process that you can analyze. Datalog generation, and datalog evaluation. In simple cases, with static authorization rules, datalog generation won't be the bottleneck, but if generating datalog involves database queries and complex generation logic, then you have optimization opportunities. The second step is datalog evaluation. There might be a balance between those two steps (ie make the datalog generation process more complex in order to simplify evaluation), so optimizations should always be considered over the whole authorization process.

Reviewer: ```suggestion
Then, there are two steps in the authorization process that you can analyze: datalog generation, and datalog evaluation. 

The first step, datalog generation, is not likely to be the bottleneck in simple cases, with static authorization rules. However, if your datalog generation involves database queries and complex generation logic, then you have optimization opportunities.  
The second step is datalog evaluation. There might be a balance between those two steps (i.e. making the datalog generation process more complex in order to simplify evaluation), so optimizations should always be considered over the whole authorization process.
```

Reviewer: Par contre c'est pas clair pour l'étape datalog evalution, qu'est-ce qu'on a à analyser ?

## Datalog performance contributors

As stated above, there are a lot of external factors that contribute to the final time and resource costs of the authorization process.

> Other things being equal, some elements in datalog code tend to have a disproportionate effect on performance. Please don't apply them without proper measurements.

Reviewer: "some elements": lesquels?

Author: les éléments listés en détail par la suite. je pourrais clarifier ça

- rules generating facts matched by other rules with will require more iterations before convergence

### Expression evaluation

> This part is implementation-dependent, advice applies primarily to the rust implementation.

Rules can contain expressions, that are evaluated after facts are matched. The biscuit specifications describes an evaluation strategy based on a stack machine, which aims at providing a fast evaluation.

#### Expensive operations

Operations on booleans, integers and dates are really simple operations and thus quite fast. Same for string equality, thanks to string interning (comparing two strings for equality is turned into an equality test on integers). Other string operations like prefix / suffix / substring tests are a bit more costly. Regex matching tends to be the worst offender, especially when there are a lot of different regexes. Regex compilation is memoized, so the cost can be amortized when one regex is used to match against several strings. However, if several regexes are matched against a single string, then the regex compilation costs will not be amortized.

#### Splitting expressions

> Expressions are tried in order. If an expression evaluates to false (or fails to evaluate), other expressions are not evaluated. By splitting simple conditions and placing them first, this allows rules to fail fast by only evaluating complex operations when needed.

Reviewer: ```suggestion
Expressions are tried in order. If an expression evaluates to false (or fails to evaluate), other expressions are not evaluated. Splitting simple conditions and placing them first allows rules to fail fast by only evaluating complex operations when needed.
```

# Authorization performance

> Authorization is likely part of your request handling hot path. As such, it is natural to try and make it as fast as possible. Signatures verification makes up for significant share of the time spent in the authorization process.

Author: agreed, initially the page was only about datalog performance and the intro paragraph explained why. this is not the case anymore

## Authorization process breakdown

The authorization process can be broken down in 4 parts:

- parsing;
- signature verification;
- datalog generation;
- datalog evaluation.

Parsing is typically one of the fastest steps; it depends only on the token size. Signature verification is where most of the time is spent; it depends mostly on the number of blocks (and their size). Datalog generation and datalog evaluation happen in tandem. That's the part where you have the most leverage. Datalog generation purely depends on how your application is designed. In many cases it can be done statically and thus have a negligible contribution to the overall runtime. Datalog evaluation depends on the actual datalog code that is evaluated.

## The first rule of performance optimization

> Don't do it

> Benchmarks done with biscuit-rust show that the whole process (parsing, signatures verification and datalog generation and evaluation) usually clocks in at around one millisecond. In a lot of cases, that will not be a bottleneck and thus not where you should work on performance optimization.

Author: this paragraph says that in most cases you should not need to work on optimizing the whole authorization process at all. The paragraph above says that, should you need to work on optimizing authorization, datalog would be the best place to do so.

That means that this section should come first (don't do it), and then the section explaining how to do it best if you have no choice but to do it anyway

## Measure

When it comes to performance optimization, the first step is always to measure things. First to determine if optimization is even needed, then to quantify progress.
This part entirely depends on your tech stack. You can start with coarse-grained traces telling you how long the whole authorization process takes, and then only dig down if optimization is needed.

> Then, there are two steps in the authorization process that you can analyze: datalog generation, and datalog evaluation. 

Author: for which it's worth it
